src/parser/Parser.ts

```diff
@@ -877,49 +886,44 @@ export class Parser {
 
             this.consumeStatementSeparators(true);
 
-            let func = new FunctionExpression({
-                parameters: params,
-                body: undefined, //body
-                functionType: functionType,
-                endFunctionType: undefined, //ending keyword
-                leftParen: leftParen,
-                rightParen: rightParen,
-                as: asToken,
-                returnTypeExpression: typeExpression
-            });
 
             //support ending the function with `end sub` OR `end function`
-            func.body = this.block();
+            let body = this.block();
             //if the parser was unable to produce a block, make an empty one so the AST makes some sense...
-            if (!func.body) {
-                func.body = new Block({
-                    statements: [],
-                    startingRange: util.createRangeFromPositions(func.range.start, func.range.start)
-                });
-            }
-            func.body.symbolTable = new SymbolTable(`Block: Function '${name?.text ?? ''}'`, () => func.getSymbolTable());
-
-            if (!func.body) {
-                this.diagnostics.push({
-                    ...DiagnosticMessages.callableBlockMissingEndKeyword(functionTypeText),
-                    range: this.peek().range
-                });
-                throw this.lastDiagnosticAsError();
-            }
 
             // consume 'end sub' or 'end function'
-            func.tokens.endFunctionType = this.advance();
+            const endFunctionType = this.advance();
             let expectedEndKind = isSub ? TokenKind.EndSub : TokenKind.EndFunction;
 
             //if `function` is ended with `end sub`, or `sub` is ended with `end function`, then
             //add an error but don't hard-fail so the AST can continue more gracefully
-            if (func.tokens.endFunctionType.kind !== expectedEndKind) {
+            if (endFunctionType.kind !== expectedEndKind) {
                 this.diagnostics.push({
-                    ...DiagnosticMessages.mismatchedEndCallableKeyword(functionTypeText, func.tokens.endFunctionType.text),
-                    range: func.tokens.endFunctionType.range
+                    ...DiagnosticMessages.mismatchedEndCallableKeyword(functionTypeText, endFunctionType.text),
+                    range: endFunctionType.range
                 });
             }
-            func.callExpressions = this.callExpressions;
+
+            if (!body) {
+                body = new Block({
+                    statements: [],
+                    startingRange: util.createBoundingRange(
+                        functionType, name, leftParen, ...params, rightParen, asToken, typeExpression, endFunctionType)
+                });
+            }
+
+            body.symbolTable = new SymbolTable(`Block: Function '${name?.text ?? ''}'`, () => func.getSymbolTable());
```

Comment:
This doesn't _need_ to be addressed now, but it might be nice to figure out how Plugins could generate this symbol table if they are creating new `Block`s themselves. Or, I guess if they're creating FunctionExpressions, maybe the inner FunctionExpression would create a SymbolTable for the block if there isn't one yet? idk, just something to consider as we get closer to the final release

Reply:
I *think* the only block that truly makes it own symbols is a function expression.... so I'll move the code that updates the symbol table to the FuncExpression constructor...

likewise for setting the `functionStatement` prop on functionExpression - I'll move it to the FunctionStatement constructor.